dissect/target/plugins/os/windows/dpapi/crypto.py

```diff
@@ -140,158 +114,135 @@ def __init_subclass__(cls):
-        raise NotImplementedError
-
-
-class __MD5HashAlgorithm(HashAlgorithm):
+class _MD5(HashAlgorithm):
     id = 0x8003
     name = "md5"
     digest_length = 128 // 8
     block_length = 512 // 8
 
 
-class __SHA1HashAlgorithm(HashAlgorithm):
+class _SHA1(HashAlgorithm):
     id = 0x8004
     name = "sha1"
     digest_length = 160 // 8
     block_length = 512 // 8
 
 
-class __HMACHashAlgorithm(__SHA1HashAlgorithm):
-    """Synonymous to SHA1"""
+class _HMAC(_SHA1):
+    """Synonymous to SHA1."""
 
     id = 0x8009
 
 
-class __SHA256HashAlgorithm(HashAlgorithm):
+class _SHA256(HashAlgorithm):
     id = 0x8004
     name = "sha256"
     digest_length = 256 // 8
     block_length = 512 // 8
 
 
-class __SHA384HashAlgorithm(HashAlgorithm):
+class _SHA384(HashAlgorithm):
     id = 0x800D
     name = "sha384"
     digest_length = 384 // 8
     block_length = 1024 // 8
 
 
-class __SHA512HashAlgorithm(HashAlgorithm):
+class _SHA512(HashAlgorithm):
     id = 0x800E
     name = "sha512"
     digest_length = 512 // 8
     block_length = 1024 // 8
 
 
-def pbkdf2(passphrase: bytes, salt: bytes, key_len: int, iterations: int, digest: Optional[str] = "sha1"):
+def pbkdf2(passphrase: bytes, salt: bytes, key_len: int, iterations: int, digest: str = "sha1") -> bytes:
     """Implementation of PBKDF2 that allows specifying digest algorithm.
-    Returns the corresponding expanded key which is key_len long.
+
+    Returns the corresponding expanded key which is ``key_len`` long.
     """
-    buff = b""
+    key = bytearray()
+
     i = 1
-    while len(buff) < key_len:
-        U = salt + struct.pack("!L", i)
+    while len(key) < key_len:
+        U = salt + i.to_bytes(4, "big")
         i += 1
+
         derived = hmac.new(passphrase, U, digestmod=digest).digest()
-        for r in range(iterations - 1):
+        for _ in range(iterations - 1):
             actual = hmac.new(passphrase, derived, digestmod=digest).digest()
-            derived = (
-                "".join([chr(int(x, 16) ^ int(y, 16)) for (x, y) in zip(derived.hex(), actual.hex())]).encode().hex()
-            )
-            result = ""
-            for j in range(len(derived)):
-                if j % 2 == 1:
-                    result += derived[j]
-            derived = bytes.fromhex(result)
-        buff += derived
-    return buff[:key_len]
-
-
-def dpapi_hmac(hash_algorithm: HashAlgorithm, pwd_hash: bytes, hmac_salt: bytes, value: bytes):
-    """Internal function used to compute HMACs of DPAPI structures"""
-    # hname = {"HMAC": "sha1"}.get(hash_algorithm.name, hash_algorithm.name) # TODO: Maybe importanbt?
-    enc_key = hmac.new(pwd_hash, digestmod=hash_algorithm.name)
-    enc_key.update(hmac_salt)
-    enc_key = enc_key.digest()
-    rv = hmac.new(enc_key, digestmod=hash_algorithm.name)
-    rv.update(value)
-    return rv.digest()
+            derived = bytes(x ^ y for x, y in zip(derived, actual))
+        key += derived
+        key.extend(derived)
```

Comment:
Did you mean for both these lines to be here?

Reply:
Also in error.